core/parser/src/main/java/com/blazebit/persistence/parser/PathTargetResolvingExpressionVisitor.java

```diff
@@ -239,6 +239,8 @@ public void visit(PropertyExpression expression) {
                     keyType = metamodel.type(typeArguments[0]);
                 }
             }
+        } else if (attribute instanceof SingularAttribute<?,?>) {
+            valueType = type = ((SingularAttribute) attribute).getType();
```

Comment:
Why do you think this is necessary? If there is some use of type variables, IMO this will break like this. `type` should already contain the same or in case of type variables the concrete type.

Reply:
This is necessary because `getJavaType()` may return the type of the identifier property within the IdClass rather than the return type of the association. We need the later, as we use this method for traversing property paths solely. We're not interested in IdClass properties. We can't even query them.

Reply:
Would you mind adding a comment about this here? Is there a test that fails if we remove this? It just seems strange because `getType` which is used to set the value on `type` should IMO produce the same result.

Reply:
It could be that this is a Hibernate specific bug, I haven't checked the spec just yet. However, I've introduced a test that fails without this change.

Reply:
After debugging this, I think this may be a Hibernate bug. Its just odd that `metamodel.entity(Entity.class). getSingularAttribute("key1").getMember()` returns `Entity.EntityId#key1` rather than `Entity#key1`. I can't find anything relevant in the JPA 2.1 spec as to why this would be useful. Nevertheless, this seems like a proper workaround. What is your worry with type variables?

It by the way is the `testBindingCteUsingNestedIdClass` test that fails without this change.

Reply:
If you look into `com.blazebit.persistence.parser.util.JpaMetamodelUtils#getIdAttributes` you can see special handling for Hibernate which behave strange when encountering a `@ManyToOne` id attribute. I think this is what might be biting you here.

Reply:
That exception should be for Hibernate 4 solely afaik? It is presented as Id attribute, but the java member and java type values just don't make sense. I've posted a question at https://discourse.hibernate.org/t/bug-in-jpa-compliance/1372/2 and 
https://gitter.im/hibernate/hibernate-orm?at=5b96b3cd33ebb72e37f9c882 , I may open a Jira ticket if I do not get some proper explanation.

What are your worries regarding type variables for this work around? Could I test whether that still works?

Reply:
I think if the resolver tries to resolve the path `e.association.key1` on a model like
```
@Entity class Base<T extends BaseAssoc> { @ManyToOne T assocation; }
@Entity class Sub1 extends Base<Assoc1> {}
@Entity class Sub2 extends Base<Assoc2> {}
@Entity class BaseAssoc {}
@Entity class Assoc1 extends BaseAssoc { Integer key1; }
@Entity class Assoc2 extends BaseAssoc { String key1; }
```

You might get weird results when using this code.

Reply:
Ill have a try and see what happens. Is such a domain already present in our test code base?

Reply:
I think `PolymorphicPropertySub1` and friends should fit that model.